### PR TITLE
eth: Add eth_blobBaseFee RPC and blob fields to eth_feeHistory

### LINKED ISSUE #29139: Add dencun `eth_blobBaseFee` and `eth_feeHistory` changes
# Rationale

These RPC changes were agree on in https://github.com/ethereum/execution-apis/pull/486, a common use case is for L2s that are trying to integrate with blobs need access to these methods.

In summary, there are two changes:

- Add `eth_blobBaseFee` which returns the expected base fee for blobs in the next block.
- Add two new blob-related fields to `eth_feeHistory`: `baseFeePerBlobGas ` and `blobGasUsedRatio`

# Implementation

I'll try to get a PR up to implement these this afternoon, I did scan and didn't see anything in flight anywhere but am happy to defer if someone else is already working on this.

### PR DESCRIPTION
**Pull request by ryanschneider**
Fixes #29139, this adds the new `eth_blobBaseFee` RPC and the new fields to `eth_feeHistory` from https://github.com/ethereum/execution-apis/pull/486.

~There's still one issue with the code: the `baseFeePerBlobGas` array is missing the "next" block so is one shorter than it should be, but I ran out of time and am not sure how much bandwidth I'll have to work on this again next week.~ Update: I believe this is working as expected now as of [cc1649a](https://github.com/ethereum/go-ethereum/pull/29140/commits/cc1649a364563651d5b18a8a3bf60ec0a17629a3).

Also, the unit test coverage could be a bit better, but I was having trouble getting the `newTestBackend` to actually raise the blob gas price even when blocks were full.  I had to add a fair amount of other logic to `newTestBackend` to support the merge and Cancun since it hasn't been touched in awhile, I borrowed as much logic from other tests as I could until it seemed to work.  

I also considered refactoring `oracle.FeeHistory` to return a struct (or at least named return values) but ultimately opted not to make too many changes w/o discussing them first.

In addition, I've tested on Sepolia:

```
$ curl -H "Content-Type: application/json" -d '{"id": 1, "jsonrpc": "2.0", "method": "eth_feeHistory", "params": ["0x5", "latest", [20,30]] }' http://$NODE:8545/ | jq .

{
  "jsonrpc": "2.0",
  "id": 1,
  "result": {
    "oldestBlock": "0x526002",
    "reward": [
      [
        "0x186a1",
        "0x186a1"
      ],
      [
        "0x186a1",
        "0x186a1"
      ],
      [
        "0x59682f00",
        "0x59682f00"
      ],
      [
        "0x0",
        "0x0"
      ],
      [
        "0x3b9aca00",
        "0x3b9aca00"
      ]
    ],
    "baseFeePerGas": [
      "0x4506f",
      "0x43d32",
      "0x44d30",
      "0x42021",
      "0x39c1d",
      "0x38cd0"
    ],
    "gasUsedRatio": [
      0.4303144,
      0.5589600333333333,
      0.3363257666666667,
      0,
      0.4337706333333333
    ],
    "baseFeePerBlobGas": [
      "0x535f4d983",
      "0x4a1bd3218",
      "0x502994693",
      "0x4d137e31b",
      "0x4741676a3",
      "0x3f5694c1f"
    ],
    "blobGasUsedRatio": [
      0.8333333333333334,
      0.3333333333333333,
      0.16666666666666666,
      0,
      1
    ]
  }
}
```

```
$ curl -H "Content-Type: application/json" -d '{"id": 1, "jsonrpc": "2.0", "method": "eth_blobBaseFee", "params": [] }' http://$NODE:8545/
{"jsonrpc":"2.0","id":1,"result":"0x3f5694c1f"}
```

### CODE REVIEW

**Comment by ryanschneider** [comment]
Also cc @lightclient since he worked on the execution spec for these changes.

File: eth/gasprice/feehistory.go [hunk 1]
```diff
@@ -316,13 +334,13 @@ func (oracle *Oracle) FeeHistory(ctx context.Context, blocks uint64, unresolvedL
...
 	if firstMissing == 0 {
-		return common.Big0, nil, nil, nil, nil
+		return common.Big0, nil, nil, nil, nil, nil, nil
 	}
 	if len(rewardPercentiles) != 0 {
 		reward = reward[:firstMissing]
 	} else {
 		reward = nil
 	}
 	baseFee, gasUsedRatio = baseFee[:firstMissing+1], gasUsedRatio[:firstMissing]
```
Comment (ryanschneider) [inline eth/gasprice/feehistory.go:344]: I think I need to do this `[:firstMissing]` slicing on `blobBaseFee` and `blobGasUsedRatio` as well, in fact the `+1` is probably what gives me the extra block.

File: eth/gasprice/feehistory.go [hunk 2]
```diff
@@ -211,9 +226,9 @@ func (oracle *Oracle) resolveBlockRange(ctx context.Context, reqEnd rpc.BlockNum
 //
 // Note: baseFee includes the next block after the newest of the returned range, because this
 // value can be derived from the newest block.
-func (oracle *Oracle) FeeHistory(ctx context.Context, blocks uint64, unresolvedLastBlock rpc.BlockNumber, rewardPercentiles []float64) (*big.Int, [][]*big.Int, []*big.Int, []float64, error) {
+func (oracle *Oracle) FeeHistory(ctx context.Context, blocks uint64, unresolvedLastBlock rpc.BlockNumber, rewardPercentiles []float64) (*big.Int, [][]*big.Int, []*big.Int, []float64, []*big.Int, []float64, error) {
```
Comment (ryanschneider) [inline eth/gasprice/feehistory.go:229]: Docstring comment on this method needs updated too.

**Comment by ryanschneider** [comment]
Oops, I broke it in [d8df733](https://github.com/ethereum/go-ethereum/pull/29140/commits/d8df733dd864e7a0cf6145f27b7765b952599ac3), but now that I need to add `nextBlobBaseFee`, I'll fix that after dinner! :)

File: eth/gasprice/feehistory.go [hunk 3]
```diff
@@ -88,6 +93,16 @@ func (oracle *Oracle) processBlock(bf *blockFees, percentiles []float64) {
...
 	}
 	bf.results.gasUsedRatio = float64(bf.header.GasUsed) / float64(bf.header.GasLimit)
+	bf.results.blobGasUsedRatio = 0.0
+	bf.results.blobBaseFee = new(big.Int)
+	if bf.header != nil && chainconfig.IsCancun(bf.header.Number, bf.header.Time) {
+		if blobGasUsed := bf.header.BlobGasUsed; blobGasUsed != nil {
+			bf.results.blobGasUsedRatio = float64(*blobGasUsed) / params.MaxBlobGasPerBlock
+		}
+		if excessBlobGas := bf.header.ExcessBlobGas; excessBlobGas != nil {
+			bf.results.blobBaseFee = eip4844.CalcBlobFee(*excessBlobGas)
```
Comment (ryanschneider) [inline eth/gasprice/feehistory.go:103]: Ok, so I think I need to look at the _previous_ block for `blobBaseFee`, then add a `nextBlobBaseFee` that should contain the value calculated from the current excess blob gas.

But currently I don't have access to the previous header in this function.

Comment (ryanschneider) [reply eth/gasprice/feehistory.go]: Added bf.parentHeader in cc1649a.

**Comment by ryanschneider** [comment]
Ok, I took another stab at it and now believe everything is working as expected, PTAL!

**Comment by lightclient** [comment]
@ryanschneider thank you for the PR. Hope you don't mind pushing some updates. I updated the logic of `eth_feeHistory` to use the current block for the blob base fee instead of the parent. I believe this is correct. Additionally, I simplified the test a bit for adding blob txs into the mix.

File: eth/gasprice/feehistory.go (hunk 3 above)
Comment (ryanschneider) [reply eth/gasprice/feehistory.go]: Oh I see, `excess_blob_gas` is actually the already computed from the parent header so ya you're right. 

> The value of excess_blob_gas can be calculated using the parent header.

File: eth/gasprice/feehistory.go [hunk 4]
```diff
@@ -78,16 +83,31 @@ type txGasAndReward struct {
...
-		bf.results.nextBaseFee = eip1559.CalcBaseFee(chainconfig, bf.header)
+	if config.IsLondon(big.NewInt(int64(bf.blockNumber + 1))) {
+		bf.results.nextBaseFee = eip1559.CalcBaseFee(config, bf.header)
 	} else {
 		bf.results.nextBaseFee = new(big.Int)
 	}
+	// Fill in blob base fee and next blob base fee.
+	if excessBlobGas := bf.header.ExcessBlobGas; excessBlobGas != nil {
+		bf.results.blobBaseFee = eip4844.CalcBlobFee(*excessBlobGas)
+		bf.results.nextBlobBaseFee = eip4844.CalcBlobFee(eip4844.CalcExcessBlobGas(*excessBlobGas, *bf.header.BlobGasUsed))
```
Comment (ryanschneider) [inline eth/gasprice/feehistory.go:100]: Ok ya this was the part that confused me. :)

**Review by ryanschneider — COMMENTED** [review]
Thanks @lightclient your changes all lgtm!

**Comment by lightclient** [comment]
Tested on sepolia: 

```json
> eth.feeHistory(5, 5435580)
{
  baseFeePerBlobGas: ["0x5dcb37612", "0x69849e63d", "0x76b5324b4", "0x69849e63d", "0x7223341d4", "0x69849e63d"],
  baseFeePerGas: ["0x8fee135", "0x88d8ddf", "0x7f86165", "0x78cbde4", "0x796c005", "0x783ad8e"],
  blobGasUsedRatio: [1, 1, 0, 0.8333333333333334, 0.16666666666666666],
  gasUsedRatio: [0.3031585666666667, 0.22748106666666668, 0.2889831, 0.5207132, 0.46073166666666665],
  oldestBlock: "0x52f0b8"
}
```

**Review by lightclient — APPROVED** [review]
LGTM

**Comment by ryanschneider** [comment]
@lightclient any updates on this PR being merged?

**Comment by lightclient** [comment]
Sorry we will merge it soon! Thank you for the bump!

**Review by fjl — APPROVED** [review]

**Pull request merged**